src/com/xilinx/rapidwright/rwroute/GlobalSignalRouting.java

```diff
@@ -0,0 +1,854 @@
+import com.xilinx.rapidwright.design.NetType;
+import com.xilinx.rapidwright.design.SiteInst;
+import com.xilinx.rapidwright.design.SitePinInst;
+import com.xilinx.rapidwright.device.ClockRegion;
+import com.xilinx.rapidwright.device.Device;
+import com.xilinx.rapidwright.device.IntentCode;
+import com.xilinx.rapidwright.device.Node;
+import com.xilinx.rapidwright.device.PIP;
+import com.xilinx.rapidwright.device.Site;
+import com.xilinx.rapidwright.device.Tile;
+import com.xilinx.rapidwright.device.Wire;
+import com.xilinx.rapidwright.placer.blockplacer.Point;
+import com.xilinx.rapidwright.placer.blockplacer.SmallestEnclosingCircle;
+import com.xilinx.rapidwright.router.RouteNode;
+import com.xilinx.rapidwright.router.RouteThruHelper;
+import com.xilinx.rapidwright.router.UltraScaleClockRouting;
+import com.xilinx.rapidwright.timing.ClkRouteTiming;
+import com.xilinx.rapidwright.util.Pair;
+
+/**
+ * A collection of methods for routing global signals, i.e. GLOBAL_CLOCK, VCC and GND.
+ * Adapted from RapidWright APIs.
+ */
+public class GlobalSignalRouting {
+	private static Design design;
+	private static RouteThruHelper routeThruHelper;
+	
+	static boolean clkDebug = false;
+	static boolean debugPrintClkPIPs = false;
+	
+	static Map<String, List<String>> crRoutes;
+	static Map<Pair<String, String>, List<Short>> bufceRowTaps;
+	static Device device;
+	
+	static Map<String, List<String>> dstINTtileRoutes;
+	static Map<String, Short> dstINTtileRoutey;
+	static ClkRouteTiming ceRouteTiming;
+	
+	public static Map<Node, RoutingNode> createdRoutingNodes;
+	static {
+		createdRoutingNodes = new HashMap<>();
+	}
+	
+	public static void setRouteMap(Map<String, List<String>> routeMap, Map<Pair<String, String>, List<Short>> delays) {
+		crRoutes = new HashMap<>();
+		crRoutes = routeMap;
+		bufceRowTaps = new HashMap<>();
+		bufceRowTaps = delays;
+	}
+	
+	public static void setCERouteTiming(ClkRouteTiming ceTiming) {
+		ceRouteTiming = ceTiming;
+		if(ceRouteTiming == null) {
+			System.err.println("CERouteTiming NOT EXIST");
+			return;
+		}
+		dstINTtileRoutes = ceRouteTiming.getDstINTtileRoute();
+	}
+	
+	public static void setDesignRoutethruHelper(Design design, RouteThruHelper routeThruHelper) {
+		GlobalSignalRouting.design = design;
+		GlobalSignalRouting.routeThruHelper = routeThruHelper;
+	}
+	
+	public static void setDebug() {
+		clkDebug = true;
+	}
+	public static void setPrintCLKPIPs() {
+		debugPrintClkPIPs = true;
+	}
+	
+	/**
+	 * Routes a clk enable net with input data.
+	 * @param clockEnable The net to be routed.
+	 * @param device The design device.
+	 */
+	public static void clkEnableRoute(Net clockEnable, Device device) {
+		Map<String, List<Node>> dstINTtilePaths = getListOfNodesFromRoutes(device, dstINTtileRoutes);
+		// Not import path after HDSTR
+		Set<PIP> ceNetPIPs = new HashSet<>();
+		Map<String, RouteNode> horDistributionLines = new HashMap<>();
+		
+		for(String intTile : dstINTtilePaths.keySet()) {
```

Comment:
When iterating over a map, it is preferable to use the `EntrySet` (see https://www.geeksforgeeks.org/difference-between-keyset-and-entryset-method-in-java-map/ for an example).  I see several instances of this throughout the code.  If you are iterating over a map and accessing both keys and values, please use the entry set.